copier/copier.go

```diff
@@ -0,0 +1,1367 @@
+	Mode            os.FileMode // dereferenced value for symlinks
+	ModTime         time.Time   // dereferenced value for symlinks
+	IsSymlink       bool
+	IsDir           bool   // dereferenced value for symlinks
+	IsRegular       bool   // dereferenced value for symlinks
+	IsArchive       bool   // dereferenced value for symlinks
+	ImmediateTarget string `json:",omitempty"` // raw link content
+}
+
+// GetResponse encodes a response for a single Get request.  It is an implementation detail of the copier package.
+type GetResponse struct {
+	Error string `json:",omitempty"`
+}
+
+// PutResponse encodes a response for a single Put request.  It is an implementation detail of the copier package.
+type PutResponse struct {
+	Error string `json:",omitempty"`
+}
+
+// StatOptions controls parts of Stat()'s behavior.
+type StatOptions struct {
+	CheckForArchives bool     // check for and populate the IsArchive bit in returned values
+	Excludes         []string // contents to pretend don't exist
+}
+
+// Stat globs the specified pattern in the specified directory and returns its
+// results.
+// If root and directory are both not specified, the current root directory is
+// used, and relative names in the globs list are treated as being relative to
+// the current working directory.
+// If root is specified and the current OS supports it, the stat() is performed
+// in a chrooted context.  If the directory is specified as an absolute path,
+// it should either be the root directory or a subdirectory of the root
+// directory.  Otherwise, the directory is treated as a path relative to the
+// root directory.
+// Relative names in the glob list are treated as being relative to the
+// directory.
+func Stat(root string, directory string, options StatOptions, globs []string) ([]*StatsForGlob, error) {
+	req := Request{
+		Request:     RequestStat,
+		Root:        root,
+		Directory:   directory,
+		Globs:       append([]string{}, globs...),
+		StatOptions: options,
+	}
+	resp, err := copier(options.Excludes, nil, nil, req)
+	if err != nil {
+		return nil, err
+	}
+	if resp.Stat.Error != "" {
+		return nil, errors.New(resp.Stat.Error)
+	}
+	return resp.Stat.Globs, nil
+}
+
+// GetOptions controls parts of Get()'s behavior.
+type GetOptions struct {
+	UIDMap, GIDMap     []idtools.IDMap // map from hostIDs to containerIDs in the output archive
+	Excludes           []string        // contents to pretend don't exist
+	ExpandArchives     bool            // extract the contents of named items that are archives
+	StripSetidBits     bool            // strip the setuid/setgid/sticky bits off of items being copied. no effect on archives being extracted
+	StripXattrs        bool            // don't record extended attributes of items being copied. no effect on archives being extracted
+	KeepDirectoryNames bool            // export directories as directories containing items rather than as the items they contain
```

Comment:
Does `don't strip the top directory's basename from the paths of items in subdirectories` read better?

As background, Dockerfile `COPY` and `ADD` have this sometimes-surprising behavior where if the source is a directory (say "src"), its contents (say "foo" and "bar") are each copied to a destination (say "dest") rather than the directory itself, producing "dest/foo" and "dest/bar" rather than "dest/src/foo" and "dest/src/bar".  This happens when the source is a glob, too, so if the glob matches both files and directories, the files and the contents of the directories are what are copied to the destination directory.